src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,273 @@
+            : html``
+      )}
+        <div class="addRow">
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
+            @click=${this._addItem}
+          ></paper-icon-button>
+          <paper-item-body>
+            <paper-input
+              id='addBox'
+              placeholder='${this.localize("ui.card.shopping-list.add_item")}'
+              @keydown=${this._addKeyPress}
+              no-label-float
+            ></paper-input>
+          </paper-item-body>
+        </div>
+        <div class="divider"></div>
+      ${repeat(
+        this._items!,
+        (item) =>
+          item.complete
+            ? html`
+        <div class="completedRow">
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
+            .itemId=${item.id}
+            @click=${this._completeItem}
+            tabindex='0'
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input
+              class="completeBox"
+              no-label-float
+              .value='${item.name}'
+              .itemId=${item.id}
+              @change=${this._saveEdit}
+            ></paper-input>
+          </paper-item-body>
+        </div>
+        `
+            : html``
+      )}
+      </ha-card>
+    `;
+  }
+
+  private renderStyle(): TemplateResult {
+    return html`
+      <style>
+        #clear {
+          float: right;
+          margin-top: -3em;
+          padding-right: 3em;
+        }
+        .addRow, .editRow, .completedRow {
+          display: flex;
+          flex-direction: row;
+        }
+        #addBox {
+          padding-left: 11px;
+        }
+        .addRow {
+          padding-left: 7px;
+        }
+        .addRow > paper-icon-button {
+          color: var(--secondary-text-color);
+        }
+        paper-icon-item {
+          border-top: 1px solid var(--divider-color);
+        }
+        paper-icon-item:first-child {
+          border-top: 0;
+        }
+        paper-checkbox {
+          padding: 11px 11px 11px 18px;
+        }
+        paper-input {
+          --paper-input-container-underline: {
+            display: none;
+          }
+          --paper-input-container-underline-focus: {
+            display: none;
+          }
+          --paper-input-container-underline-disabled: {
+            display: none;
+          }
+          position: relative;
+          top: 1px;
+        }
+        .divider {
+          height: 1px;
+          background-color: var(--divider-color);
+          margin: 10px;
+        }
+      </style>
+    `;
+  }
+
+  private async _fetchData(): Promise<void> {
+    if (this.hass) {
+      this._items = await fetchItems(this.hass);
+    }
+  }
+
+  private _completeItem(ev): void {
+    completeItem(this.hass!, ev.target.itemId, ev.target.checked).catch(() =>
+      this._fetchData()
+    );
+  }
+
+  private _addItem(ev): void {
+    if (this.shadowRoot!.querySelector("#addBox")) {
```

Comment:
Don't ever write IFs that span whole or most of a function. Convert those to a guard clause:

```ts
const addBox = this.shadowRoot!.querySelector("#addBox");

if (!addBox) { return; }

…
```